geomesa-hbase/geomesa-hbase-datastore/src/main/java/org/locationtech/geomesa/hbase/coprocessor/KryoLazyDensityCoprocessor.java

```diff
@@ -0,0 +1,110 @@
+import java.util.List;
+import java.util.Map;
+
+public class KryoLazyDensityCoprocessor extends KryoLazyDensityService implements Coprocessor, CoprocessorService {
+
+    private RegionCoprocessorEnvironment env;
+
+    @Override
+    public void start(CoprocessorEnvironment env) throws IOException {
+        if (env instanceof RegionCoprocessorEnvironment) {
+            this.env = (RegionCoprocessorEnvironment) env;
+        } else {
+            throw new CoprocessorException("Must be loaded on a table region!");
+        }
+    }
+
+    @Override
+    public void stop(CoprocessorEnvironment coprocessorEnvironment) throws IOException { }
+
+    @Override
+    public Service getService() {
+        return this;
+    }
+
+    @Override
+    public void getDensity(RpcController controller, DensityRequest request, RpcCallback<DensityResponse> done) {
+        SimpleFeatureType outputsft = SimpleFeatureTypes.createType("result", "mapkey:string,weight:java.lang.Double");
+        KryoFeatureSerializer output_serializer = new KryoFeatureSerializer(outputsft, SerializationOptions.withoutId());
+        KryoLazyDensityFilter filter = new KryoLazyDensityFilter();
+        Scan scan = new Scan();
+        scan.setFilter(filter);
+        DensityResponse response = null;
+        InternalScanner scanner = null;
+        try {
+            scanner = env.getRegion().getScanner(scan);
+            List<Cell> results = new ArrayList();
+            boolean hasMore = false;
+            Map<Tuple2<Double, Double>, Double> resultMap = new HashMap<>();
+            do {
+                hasMore = scanner.next(results);
+                for (Cell cell : results) {
+                    byte[] row  = CellUtil.cloneRow(cell);
+                    byte[] encodedSF = CellUtil.cloneValue(cell);
+                    SimpleFeature sf = output_serializer.deserialize(encodedSF);
+                    String str = (String) sf.getAttribute("mapkey");
+                    Tuple2<Integer, Integer> keyTemp = (Tuple2<Integer, Integer>) filter.deserializeParameters(str);
+                    Tuple2<Double, Double> key = filter.decodeKey(keyTemp);
+                    Double value = (Double) sf.getAttribute("weight");
+                    resultMap.put(key, resultMap.getOrDefault(key, 0.0) + value);
+                }
+                results.clear();
+            } while (hasMore);
+            List<Pair>  pairs = new ArrayList<>();
+
+            resultMap.forEach((Tuple2<Double, Double> key, Double val) -> {
+                Pair pair = Pair.newBuilder().setKey(key.toString()).setValue(val).build();
+                pairs.add(pair);
+            });
+
+            response = DensityResponse.newBuilder().addAllPairs(pairs).build();
+        } catch (IOException ioe) {
+            ResponseConverter.setControllerException(controller, ioe);
+        } catch (ClassNotFoundException cfe) {
+            cfe.printStackTrace();
+        } finally {
+            if (scanner != null) {
```

Comment:
if you're only concerned about null and IOE you can use ```IOUtils.closeQuietly(scanner)``` if its on the classpath already